### PR TITLE
Zkx 1784 handle liquidation flow when user closes liquidatable position

### PR DESCRIPTION
**Pull request by merkle-groot**
Changes:
1. Adjusts the del/liq size instead of giving an error if a larger size is passed
2. Adds direction to abr_transfer event in AccountManager
3. Fixes an edge case in liquidation flow
4. Fixes is_final and simplifies the logic

### CODE REVIEW

File: L2/contracts/Trading.cairo [hunk 1]
```diff
@@ -2941,9 +2740,50 @@ func process_and_execute_orders_recurse{
...
+                tempvar syscall_ptr = syscall_ptr;
+                tempvar range_check_ptr = range_check_ptr;
+            } else {
+                if (liq_position.market_id == market_id_) {
+                    if (liq_position.direction == [request_list_].direction) {
+                        let (new_amount_to_be_sold) = Math64x61_sub(
+                            liq_position.amount_to_be_sold, quantity_to_execute
+                        );
+
+                        let (is_below_zero_amount) = Math64x61_is_le(
```
Comment (swathireddy26) [inline L2/contracts/Trading.cairo:2755]: Shouldn't this be `Math64x61_is_equal`?

Comment (merkle-groot) [reply L2/contracts/Trading.cairo:2755]: If it's a close order and the liquidatable position is expecting a deleveraging, in this case `position_size` will be greater than the `liq_position.amount_to_be_sold`. If the user tries to close a size greater than the `liq_position.amount_to_be_sold`, the size will go negative.

To handle this case, I'm using is_le.

File: L2/contracts/Trading.cairo [hunk 2]
```diff
@@ -302,75 +308,91 @@ func execute_batch{
...
-    local quantity_remaining;
-    assert order_id = request_.order_id;
-    assert quantity_remaining = quantity_remaining_;
-
     // If it's a sell order, calculate the amount that can be executed
     if (request_.side == SELL) {
-        let (cmp_res) = Math64x61_is_le(
-            quantity_to_execute, position_details_.position_size, step_precision_
-        );
+        if (is_le(LIQUIDATION_ORDER, request_.order_type) == 1) {
```
[the remark this answers is not in the document]
Comment (merkle-groot) [reply L2/contracts/Trading.cairo:335]: Yep

File: L2/contracts/Trading.cairo [hunk 3]
```diff
@@ -302,75 +308,91 @@ func execute_batch{
...
+
+            // Return if the order is fully closed
+            let (is_zero_quantity_to_execute) = Math64x61_is_equal(
+                quantity_to_execute_liq, 0, step_precision_
+            );
 
-        if (is_zero_quantity_to_execute_final == TRUE) {
-            assert error_code = 524;
+            if (is_zero_quantity_to_execute == TRUE) {
+                assert error_code = 531;
```
[the remark this answers is not in the document]
Comment (merkle-groot) [reply L2/contracts/Trading.cairo:360]: Alright

**Review by vishnuc77 — APPROVED** [review]
Looks good

File: L2/contracts/Trading.cairo (hunk 1 above)
Comment (swathireddy26) [reply L2/contracts/Trading.cairo:2755]: Got it.

**Review by swathireddy26 — APPROVED** [review]
Nice Work!!

**Pull request merged**